### DISCUSSION TITLE
What's the best way to get the content of a plugin? i.e. for rendering the sidebar on a custom page

### DISCUSSION THREAD

**Discussion by robert-j-webb** [facebook/docusaurus#9070] [Q&A] [answered]
I want to render the sidebar on a custom page, specifically the one that comes from `docusaurus-plugin-content-docs`

`docusaurus-plugin-content-docs` does use SetGlobalData to put something like a sidebar in there, but for whatever reason, it ends up being just one line item: 

![image](https://github.com/facebook/docusaurus/assets/1647322/f5ecc8b3-4049-459c-9d38-91464c99d60e)

Additionally, it's formatted differently than what the builtins consume. 

However, under the content tab, it is present:

![image](https://github.com/facebook/docusaurus/assets/1647322/24baa68a-8673-429e-a702-9398e6395176)

Using that data, I can render a sidebar like this

```tsx
import React from 'react';
import {
  DocsSidebarProvider,
  DocsVersionProvider,
} from '@docusaurus/theme-common/internal';
import DocPageLayout from '@theme/DocPage/Layout';

export default function NotebookPage(props) {
  const versionMetadata = props.content.docsCurrentMetadata;
  const sidebarName = 'defaultSidebar';
  const sidebarItems = versionMetadata.docsSidebars[sidebarName];
  return (
    <DocsVersionProvider version={versionMetadata}>
      <DocsSidebarProvider name={sidebarName} items={sidebarItems}>
        <DocPageLayout>
          <div dangerouslySetInnerHTML={{ __html: props.content.body }}></div>
        </DocPageLayout>
      </DocsSidebarProvider>
    </DocsVersionProvider>
  );
}
```

And it looks quite right. However, the only way I could find to get that data is to literally read the JSON in the `.docusaurus` folder in my plugin. which doesn't work on first build because the JSON isn't present when my plugin runs.

Is there a better way to render the sidebar that I am unaware of? Is there a bug in my code possibly that would make global data have only one line item for the sidebar? 

By the way the overall goal which I've met (with some caveats) is to include Juypter notebooks in Docusaurus

**Comment by robert-j-webb** [comment]
Just to verify the global data limitation, I ran 

`npx create-docusaurus@latest my-website classic`

And then viewed the global data route:

![Screenshot 2023-06-14 at 12 33 03 PM](https://github.com/facebook/docusaurus/assets/1647322/4b4e0fa9-5db8-4ca8-9031-720839960a09)

Looks like for whatever reason it only gets the first item of the sidebar... Probably a bug in the plugin. However, it also returns it in a format that would require conversion so I'd rather not use it if possible.

It looks like the bug is in [globalData.ts](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-plugin-content-docs/src/globalData.ts#L49) uses [Lodash.mapValues](https://lodash.com/docs/4.17.15#mapValues) which doesn't work on sidebar, because it is not a flat object, instead it's highly nested.

**Comment by slorber** [comment, marked as answer]
Hi

We don't put too much data on the global data because this data has to be loaded to render any page of your site, so making it too big would make all your pages slower, not just the page you plan to read that data from. Heavy data structures like the sidebar are only injected in pages that render them (ie docs pages)

Currently Docusaurus is not super flexible in terms of how you can inject data from one page/plugin to another. Plugins don't see each others and are decoupled so it's not possible for example to have the docs sidebar being injected into a route created by the page plugin.

With React Server Components (Docusaurus v4? not soon 😅 ), we'll make this much more flexible and let you pick/query the data you need inside a React component, that will be passed down the tree. With swizzling you'd be able to override the default theme querying and pick extra data you eventually need anywhere in your React tree.

Related issues:
- https://github.com/facebook/docusaurus/discussions/8758
- https://github.com/facebook/docusaurus/issues/8886
- https://github.com/facebook/docusaurus/discussions/8931

**Comment by robert-j-webb** [comment]
Haha thanks! RSC + Docusarus sounds pretty interesting, would definitely make it easy to get rid of the global data. 


I figure I'll just fork to change the plugins global data. If you'd like to see it merged, I can make it an optional with default turned off, but I totally get not wanting to add more options to the docs plugin.

**Reply by slorber** [reply]
considering RSCs will bring huge flexibility in the future I'd prefer not introducing options that we'd want to get rid of quite soon

**Reply by robert-j-webb** [reply]
Makes sense. Can't wait for the RSC update, sounds really interesting.
